utils/result.hh

```diff
@@ -44,4 +44,15 @@ concept ExceptionContainerResult = bo::is_basic_result<R>::value && ExceptionCon
 template<typename F>
 concept ExceptionContainerResultFuture = seastar::is_future<F>::value && ExceptionContainerResult<typename F::value_type>;
 
+template<typename L, typename R>
+concept ResultCompatible =
```

Comment:
In what way are they compatible if value types can be completely different.

Reply:
They are compatible in the sense that they can hold the same error type. In a single piece of code you may want to use results which hold different value types but the same error type (e.g. `coordinator_result<T>` for different T's). This is useful because if two results have the same error type, you are sure you can convert one failed result to another.

Example:

```c++
template<typename T> using result = bo::result<T, unfunny_joke_error>;

result<pen> get_pen();
result<apple> get_apple();
result<apple_pen> make_apple_pen(apple a, pen p);

result<apple_pen> get_apple_pen() {
    result<pen> pen = get_pen();
    if (!pen) {
        // Conversions go like this:
        // result<pen> -> bo::failure_type<unfunny_joke_error> -> result<apple_pen>
        return pen.as_faliure();
    }
    result<apple> apple = get_apple();
    if (!apple) {
        return apple.as_failure();
    }
    return make_apple_pen(pen, apple);
}

```

Reply:
But for that types they are holding should be convertible as well. I do not think the concept checks it.

Reply:
By "ResultCompatible" here I meant compatibility of the error types and not necessarily of values. Maybe the name is just confusing and suggests too much? How about `ResultWithCompatibleError`?

If by "types they are holding" you meant the error type, then yes, the concept is too restrictive. You could convert a failed `bo::result<T, E1>` to `bo::result<U, E2>` if the error type E1 converts to E2, but the concept does not allow it at the moment. I can relax it, although there is no real benefit of doing it now because we are only using `error_container`s now and they do not support conversions between each other yet.

Reply:
> By "ResultCompatible" here I meant compatibility of the error types and not necessarily of values. Maybe the name is just confusing and suggests too much? 

It probably does since Avi also suggested to change it below.

> If by "types they are holding" you meant the error type, then yes, the concept is too restrictive. You could convert a failed bo::result<T, E1> to bo::result<U, E2> if the error type E1 converts to E2, but the concept does not allow it at the moment. I can relax it, although there is no real benefit of doing it now because we are only using error_containers now and they do not support conversions between each other yet.

No I meant that `bo::is_basic_result<L>::value` is not convertable to `bo::is_basic_result<R>::value`. So in this case how it is useful to know that error types are the same. You example will not work, no?

Reply:
I believe the example will work (after my ninja-edit, at least).

The `bo::result<T, E>` is a bit similar to `std::variant<T, E>`. If you know that your `std::variant<T, E>` holds an `E`, you can take it out, construct a `std::variant<U, E>` and return it.

It's similar for `bo::result<T, E>` -> `bo::result<U, E>`. Notice that I'm doing the `!apple` and `!pen` checks so that I'm sure that the results are holding an error. The `as_failure()` method is a type sugar which allows for the conversion to a compatible result type.